File: src/main/java/org/jukeboxmc/block/behavior/BlockBambooSapling.java
package org.jukeboxmc.block.behavior;

import org.cloudburstmc.nbt.NbtMap;
import org.cloudburstmc.protocol.bedrock.data.LevelEvent;
import org.jukeboxmc.Server;
import org.jukeboxmc.block.Block;
import org.jukeboxmc.block.BlockType;
import org.jukeboxmc.block.data.BambooLeafSize;
import org.jukeboxmc.block.data.UpdateReason;
import org.jukeboxmc.block.direction.BlockFace;
import org.jukeboxmc.event.block.BlockGrowEvent;
import org.jukeboxmc.item.Item;
import org.jukeboxmc.item.ItemType;
import org.jukeboxmc.math.Location;
import org.jukeboxmc.math.Vector;
import org.jukeboxmc.player.Player;
import org.jukeboxmc.util.Identifier;
import org.jukeboxmc.world.World;

import java.util.concurrent.ThreadLocalRandom;

public class BlockBambooSapling extends Block {

    public BlockBambooSapling(Identifier identifier) {
        super(identifier);
    }

    public BlockBambooSapling(Identifier identifier, NbtMap blockStates) {
        super(identifier, blockStates);
    }

    @Override
    public boolean placeBlock(Player player, World world, Vector blockPosition, Vector placePosition, Vector clickedPosition, Item itemInHand, BlockFace blockFace) {
        Block blockDown = this.getSide(BlockFace.DOWN);
        if (!blockDown.getType().equals(BlockType.GRASS) &&
                !blockDown.getType().equals(BlockType.DIRT) &&
                !blockDown.getType().equals(BlockType.SAND) &&
                !blockDown.getType().equals(BlockType.GRAVEL) &&
                !blockDown.getType().equals(BlockType.PODZOL)) {
            return false;
        }

        if (world.getBlock(placePosition) instanceof BlockWater || world.getBlock(placePosition, 1) instanceof BlockWater) {
            return false;
        }
        world.setBlock(placePosition, this);
        return true;
    }

    @Override
    public boolean interact(Player player, Vector blockPosition, Vector clickedPosition, BlockFace blockFace, Item itemInHand) {
        if (itemInHand.getType().equals(ItemType.BONE_MEAL)) {
            boolean success = false;
            Block block = this.getSide(BlockFace.UP);
            if (block.getType().equals(BlockType.AIR)) {
                success = this.grow(block);
            }

            if (success) {
                itemInHand.updateItem(player, 1);
                this.location.getWorld().spawnParticle( LevelEvent.PARTICLE_CROP_GROWTH, this.location );
            }
            return true;
        }
        return false;
    }

    @Override
    public long onUpdate(UpdateReason updateReason) {
        if (updateReason.equals(UpdateReason.NORMAL)) {
            Block blockDown = this.getSide(BlockFace.DOWN);
            if (!blockDown.getType().equals(BlockType.GRASS) &&
                    !blockDown.getType().equals(BlockType.DIRT) &&
                    !blockDown.getType().equals(BlockType.SAND) &&
                    !blockDown.getType().equals(BlockType.GRAVEL) &&
                    !blockDown.getType().equals(BlockType.PODZOL)) {
                this.breakNaturally();
            } else {
                Block blockUp = this.getSide(BlockFace.UP);
                if (blockUp.getType().equals(BlockType.BAMBOO)) {
                    BlockBamboo blockBambooUp = (BlockBamboo) blockUp;
                    BlockBamboo blockBamboo = Block.create(BlockType.BAMBOO);
                    blockBamboo.setBambooStalkThickness(blockBambooUp.getBambooStalkThickness());
                    this.location.getWorld().setBlock(this.location, blockBamboo);
                }
            }
            return 0;
        } else if (updateReason.equals(UpdateReason.RANDOM)) {
            Block blockUp = this.getSide(BlockFace.UP);
            if (!this.hasAge() && blockUp.getType().equals(BlockType.AIR) && ThreadLocalRandom.current().nextInt(3) == 0) {
                BlockBamboo blockBamboo = Block.create(BlockType.BAMBOO);
                blockBamboo.setBambooLeafSize(BambooLeafSize.SMALL_LEAVES, false);
                BlockGrowEvent blockGrowEvent = new BlockGrowEvent(blockUp, blockBamboo);
                Server.getInstance().getPluginManager().callEvent(blockGrowEvent);
                if (!blockGrowEvent.isCancelled()) {
                    BlockBamboo block = (BlockBamboo) blockGrowEvent.getNewState();
                    block.setLocation(new Location(this.location.getWorld(), this.location.getBlockX(), blockUp.getLocation().getBlockY(), this.location.getBlockZ()));
                    block.placeBlock(null, this.location.getWorld(), null, block.getLocation(), new Vector(0.5f, 0.5f, 0.5f), Item.create(ItemType.AIR), BlockFace.DOWN);
                }
            }
        }
        return 0;
    }

    public boolean grow(Block block) {
        BlockBamboo blockBamboo = Block.create(BlockType.BAMBOO);
        blockBamboo.setLocation(this.location);
        return blockBamboo.grow(block);
    }

    public void setAge(boolean value) {
        this.setState("age_bit", value ? (byte) 1 : (byte) 0);
    }

    public void setAge(boolean value, boolean update) {
        this.setState("age_bit", value ? (byte) 1 : (byte) 0, update);
    }

    public boolean hasAge() {
        return this.stateExists("age_bit") && this.getByteState("age_bit") == 1;
    }
}

File: JukeboxMC-Server/src/main/kotlin/org/jukeboxmc/server/block/behavior/BlockBambooSapling.kt
package org.jukeboxmc.server.block.behavior

import org.cloudburstmc.nbt.NbtMap
import org.jukeboxmc.api.Identifier
import org.jukeboxmc.api.block.BambooSapling
import org.jukeboxmc.api.block.Block
import org.jukeboxmc.api.block.BlockType
import org.jukeboxmc.api.block.data.BambooLeafSize
import org.jukeboxmc.api.block.data.BlockFace
import org.jukeboxmc.api.block.data.SaplingType
import org.jukeboxmc.api.event.block.BlockGrowEvent
import org.jukeboxmc.api.item.Item
import org.jukeboxmc.api.item.ItemType
import org.jukeboxmc.api.math.Location
import org.jukeboxmc.api.math.Vector
import org.jukeboxmc.api.world.Particle
import org.jukeboxmc.server.JukeboxServer
import org.jukeboxmc.server.UpdateReason
import org.jukeboxmc.server.block.JukeboxBlock
import org.jukeboxmc.server.extensions.isOneOf
import org.jukeboxmc.server.extensions.toByte
import org.jukeboxmc.server.extensions.toJukeboxItem
import org.jukeboxmc.server.extensions.toJukeboxPlayer
import org.jukeboxmc.server.item.JukeboxItem
import org.jukeboxmc.server.player.JukeboxPlayer
import org.jukeboxmc.server.world.JukeboxWorld
import java.util.concurrent.ThreadLocalRandom

class BlockBambooSapling(identifier: Identifier, blockStates: NbtMap?) : JukeboxBlock(identifier, blockStates),
    BambooSapling {

    override fun placeBlock(
        player: JukeboxPlayer,
        world: JukeboxWorld,
        blockPosition: Vector,
        placePosition: Vector,
        clickedPosition: Vector,
        itemInHand: JukeboxItem,
        blockFace: BlockFace
    ): Boolean {
        val blockDown = this.getRelative(BlockFace.DOWN)
        if (!blockDown.getType().isOneOf(
                BlockType.GRASS,
                BlockType.DIRT,
                BlockType.SAND,
                BlockType.GRASS,
                BlockType.PODZOL
            )
        ) {
            return false
        }
        return super.placeBlock(player, world, blockPosition, placePosition, clickedPosition, itemInHand, blockFace)
    }

    override fun interact(
        player: JukeboxPlayer,
        world: JukeboxWorld,
        blockPosition: Vector,
        clickedPosition: Vector,
        blockFace: BlockFace,
        itemInHand: JukeboxItem
    ): Boolean {
        if (itemInHand.getType() == ItemType.BONE_MEAL) {
            var success = false
            val block = this.getRelative(BlockFace.UP)
            if (block.getType() == BlockType.AIR) {
                success = this.grow(block)
            }
            if (success) {
                //TODO itemInHand.updateItem(player, 1)
                this.getWorld().spawnParticle(Particle.PARTICLE_CROP_GROWTH, this.getLocation())
            }
            return true
        }
        return false
    }

    override fun onUpdate(updateReason: UpdateReason): Long {
        if (updateReason == UpdateReason.NORMAL) {
            val blockDown = this.getRelative(BlockFace.DOWN)
            if (!blockDown.getType().isOneOf(
                    BlockType.GRASS,
                    BlockType.DIRT,
                    BlockType.SAND,
                    BlockType.GRAVEL,
                    BlockType.PODZOL
                )
            ) {
                breakNaturally()
            } else {
                val blockUp = this.getRelative(BlockFace.UP)
                if (blockUp.getType() == BlockType.BAMBOO) {
                    val blockBambooUp = blockUp as BlockBamboo
                    val blockBamboo = Block.create(BlockType.BAMBOO) as BlockBamboo
                    blockBamboo.setBambooStalkThickness(blockBambooUp.getBambooStalkThickness())
                    this.getWorld().setBlock(this.getLocation(), blockBamboo)
                }
            }
            return 0
        } else if (updateReason == UpdateReason.RANDOM) {
            val blockUp = this.getRelative(BlockFace.UP)
            if (!hasAge() && blockUp.getType() == BlockType.AIR && ThreadLocalRandom.current().nextInt(3) == 0) {
                val blockBamboo = Block.create(BlockType.BAMBOO) as BlockBamboo
                blockBamboo.setBambooLeafSize(BambooLeafSize.SMALL_LEAVES)
                val blockGrowEvent = BlockGrowEvent(blockUp, blockBamboo)
                JukeboxServer.getInstance().getPluginManager().callEvent(blockGrowEvent)
                if (!blockGrowEvent.isCancelled()) {
                    val block = blockGrowEvent.getNewState() as BlockBamboo
                    block.setLocation(
                        Location(
                            this.getWorld(),
                            this.getBlockX(),
                            blockUp.getBlockY(),
                            this.getBlockZ()
                        )
                    )
                    block.placeBlock(
                        JukeboxServer.getInstance().getPlayer("null")!!.toJukeboxPlayer(),
                        this.getWorld(),
                        Vector(0, 0, 0),
                        block.getLocation(),
                        Vector(0.5f, 0.5f, 0.5f),
                        Item.create(ItemType.AIR).toJukeboxItem(),
                        BlockFace.DOWN
                    )
                }
            }
        }
        return 0
    }

    override fun canPassThrough(): Boolean {
        return true
    }

    fun grow(block: Block): Boolean {
        val blockBamboo: BlockBamboo = Block.create<BlockBamboo>(BlockType.BAMBOO)
        blockBamboo.setLocation(this.getLocation())
        return blockBamboo.grow(block)
    }

    override fun hasAge(): Boolean {
        return this.getBooleanState("age_bit")
    }

    override fun setAge(value: Boolean): BlockBambooSapling {
        return this.setState("age_bit", value.toByte())
    }

    override fun getSaplingType(): SaplingType {
        return SaplingType.valueOf(this.getStringState("sapling_type"))
    }

    override fun setSaplingType(value: SaplingType): BlockBambooSapling {
        return this.setState("sapling_type", value.name.lowercase())
    }
}